File: MAZUET_optimisation.py
import random

def fonction(a,b,c,d,e,f,x,y):
	return a*x**2+b*y**2+c*x*y+d*x+e*y+f


a=float(input("choix de a : "))
b=float(input("choix de b : "))
c=float(input("choix de c : "))
d=float(input("choix de d : "))
e=float(input("choix de e : "))
f=float(input("choix de f : "))

xmin=float(input("choix de xmin: "))
xmax=float(input("choix de xmax: "))
while (xmin>xmax):
	print("bornes de x non conformes ! xmin<=xmax")
	xmin=float(input("choix de xmin: "))
	xmax=float(input("choix de xmax: "))

ymin=float(input("choix de ymin: "))
ymax=float(input("choix de ymax: "))
while (ymin>ymax):
	print("bornes de y non conformes ! ymin<=ymax");
	ymin=float(input("choix de ymin: "))
	ymax=float(input("choix de ymax: "))


print("Optimisation de la fonction f(x,y)=",a,"x^2+",b,"y^2+",c,"xy+",d,"x+",e,"y+",f)
n=int(input("nombre de répétitions (0<n<=100000) : ")) #maximum choisis arbitrairement
while(n<0 or n>100000):
	print("nombre de répétitions trop important !")
	n=int(input("nombre de répétitions (0<n<=100000) : "))

x_optimise = random.uniform(xmin, xmax)
y_optimise=random.uniform(ymin, ymax)
f_optimise=fonction(a,b,c,d,e,f,x_optimise,y_optimise)
if(n<=20):
	print("f(",x_optimise,",",y_optimise,")=",f_optimise)

for i in range(n-1):
	x2 = random.uniform(xmin, xmax)
	y2=random.uniform(ymin, ymax)
	f2 = fonction(a,b,c,d,e,f,x2,y2)
	if (f2<f_optimise):
		x_optimise=x2;
		y_optimise=y2;
		f_optimise=f2;
	if(n<=20):
		print("f(",x2,",",y2,")=",f2)
print("le couple pour lequel la fonction est la plus optimale est (",x_optimise,",",y_optimise,")")

File: MAZUET_optimisation.c
#include <math.h>
#include <stdlib.h>

float fonction(float a, float b, float c, float d, float e, float f, float x,
               float y);

int main() {
  float a, b, c, d, e, f, xmin, xmax, ymin, ymax; // initialisation des
                                                  // variables
  int n;
  long int u;
  time(&u);
  srand(u);

  printf("choix de a : ");
  scanf("%f", &a);
  printf("choix de b : ");
  scanf("%f", &b);
  printf("choix de c : ");
  scanf("%f", &c);
  printf("choix de d : ");
  scanf("%f", &d);
  printf("choix de e : ");
  scanf("%f", &e);
  printf("choix de f : ");
  scanf("%f", &f);

  printf("choix de xmin : ");
  scanf("%f", &xmin);
  printf("choix de xmax : ");
  scanf("%f", &xmax);
  while (xmin > xmax) {
    printf("bornes de x non conformes ! xmin<=xmax\n");
    printf("choix de xmin : ");
    scanf("%f", &xmin);
    printf("choix de xmax : ");
    scanf("%f", &xmax);
  }

  printf("choix de ymin : ");
  scanf("%f", &ymin);
  printf("choix de ymax : ");
  scanf("%f", &ymax);
  while (ymin > ymax) {
    printf("bornes de y non conformes ! ymin<=ymax\n");
    printf("choix de ymin : ");
    scanf("%f", &ymin);
    printf("choix de ymax : ");
    scanf("%f", &ymax);
  }

  printf("Optimisation de la fonction f(x,y)=%.3fx^2+%.3fy^2+%.3fxy+%.3fx+%.3fy+%.3f\n",a, b, c, d, e, f); // faciliter la compréhension en limitant le nombre de chiffres après la virgule

  printf("nombre de répétition (0<n<=100000) : "); // maximum choisis
                                                   // arbitrairement
  scanf("%d", &n);
  while (n < 0 || n > 100000) {
    printf(
        "nombre de répétition trop important ! entrer n (0<n<=100000) : "); // maximum choisis arbitrairement
    scanf("%d", &n);
  }

  float x_optimise, y_optimise, f_optimise; // plus petite valeur f(x,y)
  x_optimise = xmin + (float)rand() / ((float)RAND_MAX / (xmax - xmin));
  y_optimise = ymin + (float)rand() / ((float)RAND_MAX / (ymax - ymin));
  f_optimise = fonction(a, b, c, d, e, f, x_optimise, y_optimise);
  if (n <= 20)
    printf("f(%f,%f)=%f\n", x_optimise, y_optimise, f_optimise);

  for (int i = 1; i < n; i++) {

    float x2, y2;
    x2 = xmin + (float)rand() / ((float)RAND_MAX / (xmax - xmin));

    y2 = ymin + (float)rand() / ((float)RAND_MAX / (ymax - ymin));

    float f2 = fonction(a, b, c, d, e, f, x2, y2);
    if (f2 < f_optimise) {
      x_optimise = x2;
      y_optimise = y2;
      f_optimise = f2;
    }
    if (n <= 20)
      printf("f(%f,%f)=%f\n", x2, y2, f2);
  }
  printf("le couple pour lequel la fonction est la plus optimale est (%f,%f)",
         x_optimise, y_optimise);
}

float fonction(float a, float b, float c, float d, float e, float f, float x,
               float y) {
  return a * pow(x, 2) + b * pow(y, 2) + c * x * y + d * x + e * y + f;
}